src/lua/msgpack.c

```diff
@@ -177,9 +177,9 @@ luamp_set_decode_extension(luamp_decode_extension_f handler)
 static void
 translate_map_key_field(struct luaL_field *field, struct mh_strnu32_t *translation)
 {
-	mh_int_t key = mh_strnu32_find_str(translation,
-					   field->sval.data,
-					   field->sval.len);
+	mh_int_t key = mh_strnu32_lua_find_str(translation,
+					       field->sval.data,
+					       field->sval.len);
```

Comment:
Purpose of using Lua hash is that if you get a string value from Lua, you can use `lua_hashstring()` to get its **already** calculated hash. Look at how it is used in `lbox_tuple_field_by_path()`. Otherwise the patch is pointless, no? You replaced one hash function with another, but you still calculate the hash on each lookup.

Reply:
As I said, I see the benefit of using Lua hash for the MsgPack object dictionary:
https://github.com/CuriousGeorgiy/tarantool/blob/a565ebc16ce4a3f332908b3b39ebc1faea3c7d9d/src/lua/msgpack.c#L883-L896

In this snippet the string's hash is cleary calculated by Lua upon `lua_rawget`.

The reason I changed the hash function of the translation table is so that one table can be used both for translation and as a dictionary for a MsgPack object. 

I also [added](https://github.com/CuriousGeorgiy/tarantool/blob/afb9f864bbe53448a217762ba82c67930cee9ed5/src/lua/msgpack.c#L246-L248) `lua_hashstring` to encoding, as you, AFAI, suggest, but I don't see what benefit it can give, since in the latter snippet the string's hash isn't previously calculated by Lua.

Reply:
Sorry, I just don't see this snippet in your branch. The only related usage I see there now (after my comment is fixed) is in `luamp_encode_with_translation_r()`.

> The reason I changed the hash function of the translation table is so that one table can be used both for translation and as a dictionary for a MsgPack object.

Maybe. But in scope of this particular PR the change was useless - you didn't use `lua_hashstring()` until I pointed it out, right? Adding Lua hash "just for future" wouldn't be of much help. Or did I misunderstand what you wanted to do? My intention was to avoid hash calculation for Lua strings.

> but I don't see what benefit it can give, since in the latter snippet the string's hash isn't previously calculated by Lua

I don't know about that other snipped (which I don't see on the branch in this PR), but in the new usage in `luamp_encode_with_translation_r()` you avoid hash calculation now. `lua_hashstring()` returns an already calculated hash, it doesn't calculate it from scratch.

Reply:
> lua_hashstring() returns an already calculated hash, it doesn't calculate it from scratch.

Why do you think so?

Reply:
You should have no problems opening it in a code editor, right? It is in the luajit repo. Let me paste it here real quick:
```
LUA_API uint32_t lua_hashstring(lua_State *L, int idx)
{
  TValue *o = index2adr(L, idx);
  lua_assert(tvisstr(o));
  GCstr *s = strV(o);
  if (! strsmart(s))
    return s->hash;
  return lua_hash(strdata(s), s->len);
}
```
As you can see, those non-smart strings have a member called `hash`, which contains an already calculated hash value. It is simply returned without any calculation. You also should have no obstacles on your way to `git blame` (also available directly on github) which reveals when and why these lua-hash helpers were introduced: https://github.com/tarantool/luajit/commit/b8587fafa36b417194b208d747e1f2be1faf19be.

Reply:
Thanks a lot for elaborating on this topic, I had no idea we have such capabilities, that's really neat!